### DISCUSSION TITLE
Unsupported media type for a third party API

### DISCUSSION THREAD

**Discussion by Quafadas** [disneystreaming/smithy4s#535] [Q&A] [answered]
I'm experimenting with using smithy to drive a third party API and getting back a 415. 

The client is constructed like this; 
```
  private def ilsopClient(
      client: Client[IO],
      token: String,
      stub: String
  ): Resource[IO, IlsopService[IO]] =
    IlsopService.simpleRestJson.clientResource(
      authMiddleware(token)(client),
      Uri.unsafeFromString(stub)
    )

```
For the sake of completeness, this is a version of the request which works, in a the VSCode REST extensions. 
```
###
POST {{url}}submissions/v1/brokers
Content-Type: application/json
Authorization: Bearer {{authToken}}

{
    "name" : "SimonsTest"
}
```
Smithy model is correspondingly simple; 
```
structure MakeBroker {    
    @required name: String
}

@http(method: "POST", uri: "/submissions/v1/brokers", code: 200)
operation CreateBroker {
  input: MakeBroker,
  output: BrokerIdInput
}

```
I get this response, which is confusing me. Is smithy attempting to stream it's body or some other wizardry perhaps? Anything obvious I'm missing here?

```
smithy4s.http.UnknownErrorResponse: status 415
```

**Comment by Baccata** [comment]
Nothing obvious at a glance. 

Could you plug a middleware on the http4s client to print the "curl view" of the request (ask the http4s for instructions) before it's sent, see whether there's any mismatch between a smithy4s-formed request and a handcrafted one, and report back ?

**Comment by daddykotex** [comment]
It's a little hard to help because:

1. we don't know what the request the generated client produces
2. we don't know what the third party is and what it is expecting

For 1, we could get some help from the http4s middleware (if you use that) to add a RequestLogger so you can tell us what the generated request looks like:

```
IlsopService.simpleRestJson.clientResource(
        RequestLogger(true, true)(authMiddleware(token)(client)),
        Uri.unsafeFromString("http://127.0.0.1:9000")
      )
```

**Reply by daddykotex** [reply]
Be careful not to include your token in a github response

**Comment by Quafadas** [comment, marked as answer]
Riiiiiiiiiiiiiiiiight. I've figured this out and it's kind of an obvious but (personally massive) gotcha in hindsight. As always - thanks for writing back.

Pop quiz - what's wrong here?
```
  private def authMiddleware(inTok: String): org.http4s.client.Middleware[IO] = { (client: Client[IO]) =>
    Client { req =>
      client.run(
        req.withHeaders(
          headers.Authorization(Credentials.Token(AuthScheme.Bearer, inTok))
        )
      )
    }
  }
```
Well, I'm going to go out on a limb and say that smithy does all the right things, and adds it's "I'm json" headers. I then come along, trample and replace (including but not limited to the content-type header) your excellent work with the call to `withHeaders`. 

Now, because the "GET" requests worked, I got thrown off looking at the middleware ... because I "knew it worked :-("... 

```
  def authMiddleware(inTok: String): org.http4s.client.Middleware[IO] = { (client: Client[IO]) =>
    Client { req =>
      client.run(
        req.withHeaders(
          req.headers ++ Headers(headers.Authorization(Credentials.Token(AuthScheme.Bearer, inTok)))
        )
      )      
    }
  }
```
And apparently, all is well with the world again... arg.

**Reply by Quafadas** [reply]
P.S. I did look for the "addHeader" method, but that seems to give me nasty looking message about implicits. I guess I'm missing an import, but that's for another day.

**Reply by daddykotex** [reply]
ahhh I've been bitten by that in the past. I'll update my auth middleware example to avoid this.

[addHeader](https://www.javadoc.io/static/org.http4s/http4s-docs_2.13/0.23.16/org/http4s/Request.html#addHeader[H](h:H)(implicitevidence$1:org.http4s.Header[H,org.http4s.Header.Recurring]):Message.this.Self) works for headers than can be added multiple time (not many), you may want [putHeader](https://www.javadoc.io/static/org.http4s/http4s-docs_2.13/0.23.16/org/http4s/Request.html#putHeaders(headers:org.http4s.Header.ToRaw*):Message.this.Self)?

**Reply by Baccata** [reply]
Happy to hear we've been good rubber ducks 😄
